src/services/ftp/backend.rs

```diff
@@ -0,0 +1,512 @@
+
+    /// set root path for ftp backend.
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        self.root = if root.is_empty() {
+            None
+        } else {
+            Some(root.to_string())
+        };
+
+        self
+    }
+
+    /// set user for ftp backend.
+    pub fn user(&mut self, user: &str) -> &mut Self {
+        self.user = if user.is_empty() {
+            None
+        } else {
+            Some(user.to_string())
+        };
+
+        self
+    }
+
+    /// set password for ftp backend.
+    pub fn password(&mut self, password: &str) -> &mut Self {
+        self.password = if password.is_empty() {
+            None
+        } else {
+            Some(password.to_string())
+        };
+
+        self
+    }
+
+    /// set tls for ftp backend.
+    pub fn tls(&mut self, tls: &str) -> &mut Self {
+        self.tls = if tls.is_empty() {
+            Some("false".to_string())
+        } else {
+            Some(tls.to_string())
+        };
+
+        self
+    }
+
+    /// Build a ftp backend.
+    pub fn build(&mut self) -> Result<Backend> {
+        info!("ftp backend build started: {:?}", &self);
+        let endpoint = match &self.endpoint {
+            None => {
+                return Err(other(BackendError::new(
+                    HashMap::new(),
+                    anyhow!("endpoint must be specified"),
+                )))
+            }
+            Some(v) => v,
+        };
+
+        let port = match &self.port {
+            None => "21".to_string(),
+            Some(v) => v.clone(),
+        };
+
+        let root = match &self.root {
+            // set default path to '/ftp'
+            None => "/ftp".to_string(),
+            Some(v) => {
+                debug_assert!(!v.is_empty());
+                let mut v = v.clone();
+                if !v.starts_with('/') {
+                    return Err(other(BackendError::new(
+                        HashMap::from([("root".to_string(), v.clone())]),
+                        anyhow!("root must start with /"),
+                    )));
+                }
+                if !v.ends_with('/') {
+                    v.push('/');
+                }
+                v
+            }
+        };
+
+        let user = match &self.user {
+            None => "".to_string(),
+            Some(v) => v.clone(),
+        };
+
+        let password = match &self.password {
+            None => "".to_string(),
+            Some(v) => v.clone(),
+        };
+
+        let credential = (user, password);
+
+        let tls = match &self.tls {
+            None => false,
+            Some(v) => {
+                if v == "true" {
+                    true
+                } else {
+                    false
+                }
+            }
+        };
+
+        info!("ftp backend finished: {:?}", &self);
+        Ok(Backend {
+            endpoint: endpoint.to_string(),
+            port,
+            root,
+            credential,
+            tls,
+        })
```

Comment:
Do we need to support the `active`? Or do you have a plan to implement them in the future?

BTW, it's OK to me not to support `active` mode as nearly no one uses it.

Reply:
For now, I don't plan to add this feature. But I will implement the supprot to `active` mode if someone demand it in the future.

Reply:
LGTM, thanks!